src/Socket.cpp

```diff
@@ -34,57 +36,54 @@ void Socket::write(const std::string& buff) {
     write(buff.c_str(), buff.size());
 }
 
-void Socket::read(char* buff, size_t buffsize) {
-    while(buffsize > 0) {
-        ssize_t res = ::read(_fd, buff, buffsize);
-        if(res == 0) {
-            throw std::runtime_error("Socket::read() - client closed connection");
-        } else if(res < 0) {
-            if(errno == EINTR)
-                continue;
-            throw std::runtime_error("Socket::read() - read() failed");
-        }
+void Socket::read(char* buff, const size_t buffsize) {
+    while(_buffer.size() < buffsize)
+        bufferedRead();
 
-        buff += res;
-        buffsize -= res;
-    }
+    strncpy(buff, _buffer.c_str(), buffsize);
+    buff[buffsize] = '\0';
+    _buffer = _buffer.substr(buffsize);
 }
 
 /*
  * Reads a line terminated by \r\n. Returns length of a received string.
- *
- * Current implementation reads bytes one by one. It's probably not the
- * best approach, but it's simple and doesn't require a temporary buffer
- * which should be somehow shifted all the time. So until this method
- * becomes a bottleneck for someone it's good enough.
  */
-size_t Socket::readLine(char* buff, size_t buffsize) {
-    size_t currsize = 0;
-    // TODO: optimize, figure out how many bytes are available using ioctl
-    while(currsize < buffsize) {
-        ssize_t res = ::read(_fd, buff + currsize, 1);
-        if(res <= 0) {
-            if(errno == EINTR)
-                continue;
+size_t Socket::readLine(char* buff, const size_t buffsize) {
+    while(_buffer.find_first_of('\n', 0) == std::string::npos)
```

Comment:
Sure, but you are doing a redundant work checking it all the time from the beginning of the line. If there is no \n symbol in the first say 300 bytes there is no need to re-check them all the time. On the next iteration you can start from the 301'st byte.